net/libfab/libfab.c

```diff
@@ -0,0 +1,2886 @@
+					libfab_rxep_comp_read(cq, xep);
+				}
+			} m0_tl_endfor;
+		}
+
+		libfab_bulk_buf_process(tm);
+		libfab_tm_buf_timeout(tm);
+		libfab_tm_buf_done(tm);
+
+		M0_ASSERT(libfab_tm_invariant(tm));
+		libfab_tm_unlock(tm);
+	}
+}
+
+/** 
+ * Converts network end-point to its libfabric structure.
+ */
+static inline struct m0_fab__ep *libfab_ep(struct m0_net_end_point *net)
+{
+	return container_of(net, struct m0_fab__ep, fep_nep);
+}
+
+/**
+ * Compares the endpoint name with the passed name string and
+ * returns true if equal, or else returns false
+ * If name is null then the ipaddress and port fields are compared to check for
+ * a matching endpoint.
+ */
+static bool libfab_ep_cmp(struct m0_fab__ep *ep, const char *name,
+			  struct m0_fab__ep_name *epname)
+{
+	bool ret = false;
+	M0_PRE(ep != NULL);
+	if (name == NULL) {
+		M0_ASSERT(epname != NULL);
+		if ((strcmp(ep->fep_name.fen_addr, epname->fen_addr) == 0) &&
+		    (strcmp(ep->fep_name.fen_port, epname->fen_port) == 0))
+			ret = true;
+	} else {
+		if (strcmp(ep->fep_name.fen_str_addr, name) == 0)
+			ret = true;
+	}
+	return ret;
+}
+
+/**
+ * Search for the ep in the existing ep list using one of the following -
+ *   1) Name in str format      OR
+ *   2) ipaddr and port 
+ * If found then return the ep structure, or else create a new endpoint
+ * with the name
+ * Returns 0 if the endpoint is found/created succesfully,
+ * else returns error code.
+ */
+static int libfab_ep_find(struct m0_net_transfer_mc *tm, const char *name,
+			  struct m0_fab__ep_name *epn,
+			  struct m0_net_end_point **epp)
+{
+	struct m0_net_end_point  *net;
+	struct m0_fab__ep        *ep;
+	struct m0_fab__active_ep *aep;
+	struct m0_fab__tm        *ma;
+	char                      ep_str[LIBFAB_ADDR_STRLEN_MAX] = {'\0'};
+	char                     *wc = NULL;
+	bool                      found = false;
+	int                       rc = 0;
+
+	M0_ASSERT(libfab_tm_is_locked(tm->ntm_xprt_private));
+	m0_tl_for(m0_nep, &tm->ntm_end_points, net) {
+		ep = libfab_ep(net);
+		if (libfab_ep_cmp(ep, name, epn)) {
+			*epp = &ep->fep_nep;
+			found = true;
+			libfab_ep_get(ep);
+			break;
+		}
+	} m0_tl_endfor;
+
+	if (found == false) {
+		if (name != NULL)
+			rc = libfab_ep_create(tm, name, epn, epp);
+		else {
+			M0_ASSERT(epn != NULL);
+			M0_ASSERT((strlen(epn->fen_addr) + strlen(epn->fen_port)
+				  + 8) < LIBFAB_ADDR_STRLEN_MAX);
+			sprintf(ep_str, "libfab:%s:%s", epn->fen_addr,
+				epn->fen_port);
+			rc = libfab_ep_create(tm, ep_str, epn, epp);
+		}
+	} else {
+		if (name != NULL && epn != NULL) {
+			wc = strchr(name,'*');
+			if (wc != NULL && 
+			    strcmp(ep->fep_name.fen_port, epn->fen_port) != 0) {
+				strcpy(ep->fep_name.fen_addr, epn->fen_addr);
+				strcpy(ep->fep_name.fen_port, epn->fen_port);
+				aep = libfab_aep_get(ep);
+				ma = tm->ntm_xprt_private;
+				if (aep->aep_tx_state == FAB_CONNECTED) {
+					libfab_txep_init(aep, ma);
+				}
+			}
+		}
+	}
+
+	return M0_RC(rc);
```

Comment:
here rc can be != 0. most likely libfab_ep_put(ep) along with our resource releasing is missing in such cases.